File: codechallenge/subseq.rb

def binary_search(lst, lower, upper, target)
  return -1 if lower > upper
  
  mid = (lower+upper)/2
  
  if (lst[mid][0] == target)
    mid
  elsif (target < lst[mid][0])
    binary_search(lst, lower, mid-1, target)
  else
    binary_search(lst, mid+1, upper, target)
  end
end

def find_indicies_less_than(sorted, x)
  ret = binary_search(sorted, 0, sorted.size - 1, x)
  while (sorted[ret +1] == sorted[ret])
    ret += 1
  end
  (0..ret).map { |i| sorted[i][1] }
end

def solve(sequence, weights)
  n = sequence.size
  dp = Array.new(n)
  dp[0] = weights[0]
  
  sorted_idx = sequence.each_with_index.map {|x, i| [x, i]}
  sorted_idx.sort! {|a, b| a[0] <=> b[0]}
  
  for i in 1.upto n-1
    dp[i] = weights[i]
    
    for j in (i-1).downto(0)
    #for j in find_indicies_less_than(sorted_idx, sequence[i])
      if (sequence[j] < sequence[i] and dp[j] + weights[i] > dp[i])
        dp[i] = dp[j] + weights[i]
      end
    end
  end
  
  dp.max
end

numtests = gets.to_i
numtests.times do
  gets
  sequence = gets.split.map { |x| x.to_i }
  weights = gets.split.map { |x| x.to_i }
  p solve(sequence, weights)
end

File: codechallenge/subseq.cpp
#include <stdio>
#include <stdlib>
#include <math>
#include <algorithm>
using namespace std;

unsigned int solve(unsigned int *sequence, unsigned int *weights, unsigned int n) {
  unsigned int *dp = new unsigned int [n];
  unsigned int *sorted = new unsigned int [n];
  unsigned int i, max = 0;
  
  dp[0] = weights[0];
  sort(
  
  for (i = 1; i < n; i++) {
    int j;
    
    dp[i] = weights[i];
    
    for (j = i - 1; j >= 0; j--) {
      if (dp[j] + weights[i] > dp[i] && sequence[j] < sequence[i]) {
        dp[i] = dp[j] + weights[i];
        if (dp[i] > max) max = dp[i];
      }
    }
  }
  
  return max;
}

int main() {
  unsigned int tests;
  scanf("%d", &tests);
  for ( ; tests > 0; tests--) {
    unsigned int n, i, *sequence, *weights;
    scanf("%d", &n);
    
    sequence = new unsigned int [n];
    weights = new unsigned int [n];
    
    for (i = 0; i < n; i++) {
      scanf("%d", sequence + i);
    }
    for (i = 0; i < n; i++) {
      scanf("%d", weights + i);
    }
    
    printf("%d\n", solve(sequence, weights, n));
    
    free(sequence);
    free(weights);
  }
  return 0;
}